### DISCUSSION TITLE
Improvement of "withDataTransformationFunction"

### DISCUSSION THREAD

**Discussion by bdzzaid** [firegloves/MemPOI#25] [Ideas]
@firegloves I need to export a custom column in the excel to allow my users to create groups in the final Excel file : 
```sql
SELECT gender, birthday, 'ageRange' as 'ageRange', 'myCachedData' as 'myCachedData'  FROM person
```
``Age range`` is a calculated result: 
```java
if (age <= 49) ageCategory = "20-49";
else if (age <= 64) ageCategory = "50-64";
else ageCategory = "65+"; 
```
In this SELECT 'ageRange' can by calculate with this ``SELECT : YEAR(CURDATE()) - YEAR(birthdate) AS age`` and then use DataTransformationFunction, no pb.

But, in this SELECT 'myCachedData' cannot be field in the SELECT and  then I would like to be able to use DataTransformationFunction to fill the column data. 

``myCachedData`` is a data loaded on server cache from other source then the database, and the result depends on the gender and the birthday date. 

So I need to populate it on runtime and I'm currently limited with the current DataTransformationFunction contract. For this reason I would like to improve DataTransformationFunction 

```
public abstract class DataTransformationFunction<I, O> {

    public abstract O execute(final Object value);
// ... 
    protected abstract O transform(final I value) throws MempoiException;
}
```

To 

```java 
public abstract class DataTransformationFunction<I, O> {

    public abstract O execute(final ResultSet rs, final Object value);
// ... 
    protected abstract O transform(final ResultSet rs, final I value) throws MempoiException;
}
```
And then, in DataStrategos.createDataRows add the resultset during the export process 
```java 
if (optDataTransformationFunction.isPresent()) {
	cellValue = optDataTransformationFunction.get().execute(value);
} 
```
To 
```java 
if (optDataTransformationFunction.isPresent()) {
	cellValue = optDataTransformationFunction.get().execute(rs, value);
} 
```
With the resultset, In parameter, I'm able to retrieve the ``birthday`` and the ``gender`` of the current processing row and then display the right cached data.

In a second part, I would like to improve the usage of DataTransformationFunction as a framework interface, I would like to change it to extact the @FunctionalInterface 
```java 
@FunctionalInterface
public interface DataTransformationFunctionnalInterface<I, O> {

    public static final String TRANSFORM_METHOD_NAME = "transform";

    default O execute(Object value) {
        return applyTransformation(value, (Class<I>)this.<I>getClass());
    }

    default O applyTransformation(final Object value, Class<I> clazz) {
        I castValue = this.<I>cast("null", value, clazz);
        return this.transform(castValue);
    }

    default <T> T cast(String colname, Object obj, Class<T> toType) throws MempoiException {
        try {
            if (obj instanceof Integer && toType.equals(Double.class)) {
                return (T) new Double(((Integer) obj).intValue());
            } else if (obj instanceof Float && toType.equals(Double.class)) {
                return (T) new Double(((Float) obj).floatValue());
            } else {
                return toType.cast(obj);
            }
        } catch (Exception e) {
            throw new MempoiException(
                    String.format(Errors.ERR_DATA_TRANSFORMATION_FUNCTION_CAST_EXCEPTION, colname, toType,
                            obj.getClass()));
        }
    }

    O transform(final I value) throws MempoiException;

}

```

Let me now if you agree these improvements.

**Comment by firegloves** [comment]
with a bit of delay, but it has been merged, without the DataTransformationFunctionnalInterface improvement
